app/src/main/kotlin/com/mapbox/vision/teaser/utils/PermissionsUtils.kt

```diff
@@ -7,46 +7,53 @@ import com.mapbox.vision.teaser.MainActivity
 
 object PermissionsUtils {
 
+    // PERMISSION_FOREGROUND_SERVICE was added for targetSdkVersion >= 28, it is normal and always granted, but should be added to the Manifest file
+    // on devices with Android < P(9) checkSelfPermission(PERMISSION_FOREGROUND_SERVICE) can return PERMISSION_DENIED, but in fact it is GRANTED, so skip it
+    // https://developer.android.com/guide/components/services#Foreground
     private const val PERMISSION_FOREGROUND_SERVICE = "android.permission.FOREGROUND_SERVICE"
+
+    // required by firebase dependencies, optional
+    private const val PERMISSION_C2M = "com.google.android.c2dm.permission.RECEIVE"
+
+    // required by firebase dependencies, optional
+    private const val PERMISSION_REFERRER = "com.google.android.finsky.permission.BIND_GET_INSTALL_REFERRER_SERVICE"
+
     private const val PERMISSIONS_REQUEST_CODE = 123
 
-    fun requestPermissionsIfNotGranted(activity: MainActivity): Boolean {
-        if (!allPermissionsGranted(activity) && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
-            activity.requestPermissions(getRequiredPermissions(activity), PERMISSIONS_REQUEST_CODE)
+    fun requestPermissions(activity: MainActivity): Boolean {
+        val notGranted = getNotGrantedPermissions(activity)
+        if (notGranted.isNotEmpty() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            activity.requestPermissions(notGranted, PERMISSIONS_REQUEST_CODE)
             return true
         }
         return false
     }
 
-    private fun allPermissionsGranted(activity: MainActivity): Boolean {
-        for (permission in getRequiredPermissions(activity)) {
-            if (ContextCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) {
-                // PERMISSION_FOREGROUND_SERVICE was added for targetSdkVersion >= 28, it is normal and always granted, but should be added to the Manifest file
-                // on devices with Android < P(9) checkSelfPermission(PERMISSION_FOREGROUND_SERVICE) can return PERMISSION_DENIED, but in fact it is GRANTED, so skip it
-                // https://developer.android.com/guide/components/services#Foreground
-                if (permission == PERMISSION_FOREGROUND_SERVICE) {
-                    continue
-                }
-                return false
-            }
+    fun arePermissionsGranted(activity: MainActivity, requestCode: Int) =
+        requestCode == PERMISSIONS_REQUEST_CODE
+        && getNotGrantedPermissions(activity).none { permission ->
```

Comment:
Is this possible that Google Services installed but  firebase\c2m\... permissions are not granted somehow?
Seems this logic doesn't handle this scenario (again, I am not sure this is valid use-case)

Reply:
It's the opposite - google services is not installed and this permissions fail to grant

Reply:
Yes, I mean is it possible when Google Services installed, but permissions are not granted for some reasons and we skip them as optional

Reply:
Well, they do not require explicit user feedback, so they are granted as soon as you request them. 
And, since we use firebase only for crashes right now - it's not critical if we won't get it. 
So it seems OK to me

Reply:
Okay, got it
You always get this permissions if google services installed. If not - permissions failed and we skips them

Reply:
Yup yup